flecsi/data/local/policy.hh

```diff
@@ -0,0 +1,460 @@
+    }
+  }
+
+  size2 size() const {
+    return s;
+  }
+
+  // The span is safe because it is used only within a user task while the
+  // vectors are resized or destroyed only outside user tasks (though perhaps
+  // during execute).
+  template<class T,
+    exec::task_processor_type_t ProcessorType =
+      exec::task_processor_type_t::loc>
+  util::span<T> get_storage(field_id_t fid) {
+    return get_storage<T, ProcessorType>(fid, s.second);
+  }
+
+  template<class T,
+    exec::task_processor_type_t ProcessorType =
+      exec::task_processor_type_t::loc>
+  util::span<T> get_storage(field_id_t fid, std::size_t nelems) {
+    auto & v = storages.get_storage(fid);
+    std::size_t nbytes = nelems * sizeof(T);
+    if(nbytes > v.size())
+      v.resize(nbytes);
+    return {reinterpret_cast<T *>(v.template data<ProcessorType>()), nelems};
+  }
+
+  decltype(auto) get_future(field_id_t fid) {
+    return storages.get_future(fid);
+  }
+
+  auto get_field_info(field_id_t fid) const {
+    for(auto f : fs) {
+      if(f->fid == fid)
+        return f;
+    }
+    throw std::runtime_error("can not find field");
+  }
+
+private:
+  size2 s; // (nranks, nelems)
+  fields fs; // fs[].fid is only unique within a region, i.e. r0.fs[].fid is
+             // unrelated to r1.fs[].fid even if they have the same value.
+
+  data::backend_storage_data storages;
+};
+
+namespace detail {
+struct region {
+  region(size2 s, const fields & fs, const char * = nullptr)
+    : p(new local::region_impl(s, fs)) {}
+
+  size2 size() const {
+    return p->size();
+  }
+
+  local::region_impl & operator*() const {
+    return *p;
+  }
+
+private:
+  // to preserve an address on move
+  std::unique_ptr<local::region_impl> p;
+};
+
+struct partition {
+  Color colors() const {
+    // number of rows, essentially the number of MPI ranks.
+    return r->size().first;
+  }
+
+  template<typename T,
+    exec::task_processor_type_t ProcessorType =
+      exec::task_processor_type_t::loc>
+  auto get_storage(field_id_t fid) const {
+    return r->template get_storage<T, ProcessorType>(fid, nelems);
+  }
+
+  decltype(auto) get_future(field_id_t fid) const {
+    return r->get_future(fid);
+  }
+
+  template<topo::single_space>
+  const partition & get_partition(field_id_t) const {
+    return *this;
+  }
+
+private:
+  local::region_impl * r;
+
+protected:
+  partition(region & r) : r(&*r) {}
+
+  local::region_impl & get_base() const {
+    return *r;
+  }
+
+  // number of elements in this partition on this particular rank.
+  size_t nelems = 0;
+};
+
+} // namespace detail
+} // namespace local
+
+// This type must be defined outside of namespace local to support
+// forward declarations
+struct partition // instead of "using partition ="
+  : local::detail::partition {
+
+  using local::detail::partition::partition;
+};
+
+namespace local {
+namespace detail {
+struct rows : data::partition {
+  explicit rows(region & r) : data::partition(r) {
```

Comment:
Inside the class an unqualified `partition` refers to the injected-class-name inherited from `data::partition`.  I wrote about templates because the injected-class-name can't be found in a dependent base:

```cpp
namespace N {
  struct A {};
}
template<class> struct A {A(int);};
namespace N {
  template<class T>
  struct B : ::A<T> {
    B(int i) : A(i) {}  // error: N::A is not a base of this class
    B() : ::A<T>(0) {}  // OK, template arguments required
    B(...) : B::A(0) {}  // OK, no template arguments
  };
}
```

Reply:
I misunderstood what you meant. I agree with your assessment.